OpenRVS/classes/OpenMultiPlayerWidget.uc

```diff
@@ -308,22 +329,219 @@ function Refresh(bool bActivatedByUser)
+			}
+			Timer.StartTimer(ServerList[j].IP);
+			OpenClientBeaconReceiver(m_GameService.m_ClientBeacon).QuerySingleServer(self,
+				Left(ServerList[j].IP,InStr(ServerList[j].IP,":")),
+				int(Mid(ServerList[j].IP,InStr(ServerList[j].IP,":")+1))+1000);
+		}
+		j++;
+		//CurServer = R6WindowListServerItem(CurServer.Next);
+	}
+}
+
+//0.8 written
+//1.3 added mod keyword locking
+//1.5 receive locked info from specific servers, not master list
+function ReceiveServerInfo(string sIP,coerce int iNumP,coerce int iMaxP,string sGMode,string sMapName,string sSvrName,string sModName,bool bSvrLocked)
+{
+	local R6WindowListServerItem CurServer;
+	local int i,iTime;
+/*
+	//0.8
+	//find the server in the list that we received info for, and update
 	CurServer = R6WindowListServerItem(m_ServerListBox.GetItemAtIndex(0));
 	while ( CurServer != none )
 	{
-		//1.5 - get rough ping
-		if ( Timer == none )
+		if ( CurServer.szIPAddr == sIP )
 		{
-			Timer = new class'OpenTimer';
-			Timer.ClockSource = GetEntryLevel();
+			CurServer.iMaxPlayers = iMaxP;
+			CurServer.iNumPlayers = iNumP;
+			CurServer.szName = sSvrName;
+			CurServer.szGameType = GetLevel().GetGameNameLocalization(sGMode);
+			CurServer.szMap = sMapName;
+			//1.3 - grey out version if not the right mod
+			if ( caps(class'Actor'.static.GetModMgr().m_pCurrentMod.m_szKeyWord) != caps(sModName) )
+				CurServer.bSameVersion = false;
+			else
+				CurServer.bSameVersion = true;
+			CurServer.bLocked = bSvrLocked;//1.5 added locked here
+			//1.5 add ping
+			i = Timer.EndTimer(CurServer.szIPAddr);
+			if ( i != -1 )
+				CurServer.iPing = i / 2;//divide by two for single-trip time - consistent with how Ubi used to measure
+			CurServer = none;//break the while loop
 		}
-		Timer.StartTimer(CurServer.szIPAddr);
-		OpenClientBeaconReceiver(m_GameService.m_ClientBeacon).QuerySingleServer(self,
-			Left(CurServer.szIPAddr,InStr(CurServer.szIPAddr,":")),
-			int(Mid(CurServer.szIPAddr,InStr(CurServer.szIPAddr,":")+1))+1000);
+		else
 		CurServer = R6WindowListServerItem(CurServer.Next);
+	}*/
+	
+	//1.5
+	//modify the servers array rather than the list items
+	//then rebuild list items with GetGSServers()
+	i = 0;
+	while ( i < OpenQueries.length )
+	{
+		if ( OpenQueries[i] == sIP )//close open query
+		{
+			OpenQueries.remove(i,1);
+			break;
+		}
+		i++;
 	}
+	i = 0;
+	while ( i < ServerList.length )
+	{
+		if ( ServerList[i].IP == sIP )
+		{
+			ServerList[i].ServerName = sSvrName;
+			ServerList[i].Locked = bSvrLocked;
+			ServerList[i].iMaxPlayers = iMaxP;
+			ServerList[i].iNumPlayers = iNumP;
+			ServerList[i].szGameType = GetLevel().GetGameNameLocalization(sGMode);
+			ServerList[i].szMap = sMapName;
+			ServerList[i].bWrongVersion = ( caps(class'Actor'.static.GetModMgr().m_pCurrentMod.m_szKeyWord) != caps(sModName) );
+			iTime = Timer.EndTimer(sIP);
+			if ( iTime != -1 )
+				ServerList[i].iPing = iTime / 2;//divide by two for single-trip time - consistent with how Ubi used to measure

```

Comment:
Added divide by 2 to rtt - for consistency with what we think Native Get Ping was doing.